Three higest bits is used (undocumented!) for Shared Bits from Win32 CreateFile call

diff -u cifs-orig148/cifs_debug.c cifs/cifs_debug.c
--- cifs-orig148/cifs_debug.c	2007-04-05 02:11:56 +0400
+++ cifs/cifs_debug.c	2007-06-05 12:24:45 +0400
@@ -456,6 +456,9 @@
 static read_proc_t linuxExtensionsEnabled_read;
 static write_proc_t linuxExtensionsEnabled_write;
 
+static read_proc_t etersoft_read;
+static write_proc_t etersoft_write;
+
 void
 cifs_proc_init(void)
 {
@@ -519,6 +522,12 @@
 	if (pde)
 		pde->write_proc = lookupFlag_write;
 
+	pde =
+	create_proc_read_entry("Etersoft", 0, proc_fs_cifs,
+				etersoft_read, NULL);
+	if (pde)
+		pde->write_proc = etersoft_write;
+
 /*	pde =
 	    create_proc_read_entry("NTLMV2Enabled", 0, proc_fs_cifs,
 				ntlmv2_enabled_read, NULL);
@@ -552,6 +561,7 @@
 	remove_proc_entry("LinuxExtensionsEnabled",proc_fs_cifs);
 	remove_proc_entry("Experimental",proc_fs_cifs);
 	remove_proc_entry("LookupCacheEnabled",proc_fs_cifs);
+	remove_proc_entry("Etersoft",proc_fs_cifs);
 	remove_proc_entry("cifs", proc_root_fs);
 }
 
@@ -981,4 +991,43 @@
 } */
 
 
+static int
+etersoft_read(char *page, char **start, off_t off,
+		     int count, int *eof, void *data)
+{
+	int len;
+
+	len = sprintf(page, "%d\n", etersoft_flag);
+
+	len -= off;
+	*start = page + off;
+
+	if (len > count)
+		len = count;
+	else
+		*eof = 1;
+
+	if (len < 0)
+		len = 0;
+
+	return len;
+}
+static int
+etersoft_write(struct file *file, const char __user *buffer,
+		      unsigned long count, void *data)
+{
+	char c;
+	int rc;
+
+	rc = get_user(c, buffer);
+	if (rc)
+		return rc;
+	if (c == '0' || c == 'n' || c == 'N')
+		etersoft_flag = 0;
+	else if (c == '1' || c == 'y' || c == 'Y')
+		etersoft_flag = 1;
+
+	return count;
+}
+
 #endif
diff -u cifs-orig148/cifsfs.c cifs/cifsfs.c
--- cifs-orig148/cifsfs.c	2007-04-05 02:11:56 +0400
+++ cifs/cifsfs.c	2007-06-05 12:26:04 +0400
@@ -60,6 +60,7 @@
 int cifsFYI = 0;
 int cifsERROR = 1;
 int traceSMB = 0;
+unsigned int etersoft_flag = 1;
 unsigned int oplockEnabled = 1;
 unsigned int experimEnabled = 0;
 unsigned int linuxExtEnabled = 1;
diff -u cifs-orig148/cifsglob.h cifs/cifsglob.h
--- cifs-orig148/cifsglob.h	2007-04-05 02:11:56 +0400
+++ cifs/cifsglob.h	2007-06-05 12:26:38 +0400
@@ -592,6 +592,8 @@
 GLOBAL_EXTERN atomic_t midCount;
 
 /* Misc globals */
+GLOBAL_EXTERN unsigned int etersoft_flag; /* if enabled allows extended shared modes in open */
+
 GLOBAL_EXTERN unsigned int multiuser_mount; /* if enabled allows new sessions
 				to be established on existing mount if we
 				have the uid/password or Kerberos credential 


diff -u cifs-orig148/cifsproto.h cifs/cifsproto.h
--- cifs-orig148/cifsproto.h	2007-04-05 02:11:56 +0400
+++ cifs/cifsproto.h	2007-05-17 15:54:32 +0400
@@ -263,7 +263,7 @@
 
 extern int CIFSSMBOpen(const int xid, struct cifsTconInfo *tcon,
 			const char *fileName, const int disposition,
-			const int access_flags, const int omode,
+			const int access_flags, const int share_flags, const int omode,
 			__u16 * netfid, int *pOplock, FILE_ALL_INFO *,
 			const struct nls_table *nls_codepage, int remap);
 extern int SMBLegacyOpen(const int xid, struct cifsTconInfo *tcon,

diff -u cifs-orig148/cifssmb.c cifs/cifssmb.c
--- cifs-orig148/cifssmb.c	2007-04-05 02:11:56 +0400
+++ cifs/cifssmb.c	2007-06-05 12:33:04 +0400
@@ -972,6 +972,11 @@
 	int name_len;
 	__u16 count;
 
+	if (etersoft_flag) {
+		printk("Etersoft: Do not use SMBLegacyOpen!\n");
+		return rc;
+	}
+
 OldOpenRetry:
 	rc = smb_init(SMB_COM_OPEN_ANDX, 15, tcon, (void **) &pSMB,
 		      (void **) &pSMBr);
@@ -1075,7 +1080,7 @@
 int
 CIFSSMBOpen(const int xid, struct cifsTconInfo *tcon,
 	    const char *fileName, const int openDisposition,
-	    const int access_flags, const int create_options, __u16 * netfid,
+	    const int access_flags, const int share_flags, const int create_options, __u16 * netfid,
 	    int *pOplock, FILE_ALL_INFO * pfile_info, 
 	    const struct nls_table *nls_codepage, int remap)
 {
@@ -1134,7 +1139,7 @@
 	/*  Above line causes problems due to vfs splitting create into two
 		pieces - need to set mode after file created not while it is
 		being created */
-	pSMB->ShareAccess = cpu_to_le32(FILE_SHARE_ALL);
+	pSMB->ShareAccess = cpu_to_le32(share_flags);
 	pSMB->CreateDisposition = cpu_to_le32(openDisposition);
 	pSMB->CreateOptions = cpu_to_le32(create_options & CREATE_OPTIONS_MASK);
 	/* BB Expirement with various impersonation levels and verify */

diff -u cifs-orig148/dir.c cifs/dir.c
--- cifs-orig148/dir.c	2007-04-05 02:11:56 +0400
+++ cifs/dir.c	2007-05-17 15:54:32 +0400
@@ -207,7 +207,7 @@
 	}
 	if (cifs_sb->tcon->ses->capabilities & CAP_NT_SMBS) 
 		rc = CIFSSMBOpen(xid, pTcon, full_path, disposition,
-			 desiredAccess, CREATE_NOT_DIR,
+			 desiredAccess, FILE_SHARE_ALL, CREATE_NOT_DIR,
 			 &fileHandle, &oplock, buf, cifs_sb->local_nls,
 			 cifs_sb->mnt_cifs_flags & CIFS_MOUNT_MAP_SPECIAL_CHR);
 	else
@@ -408,6 +408,7 @@
 					  WRITE_OWNER | WRITE_DAC be better? */,
 					 /* Create a file and set the
 					    file attribute to SYSTEM */
+					FILE_SHARE_ALL,
 					 CREATE_NOT_DIR | CREATE_OPTION_SPECIAL,
 					 &fileHandle, &oplock, buf,
 					 cifs_sb->local_nls,

diff -u cifs-orig148/fcntl.c cifs/fcntl.c
--- cifs-orig148/fcntl.c	2007-04-05 02:11:56 +0400
+++ cifs/fcntl.c	2007-05-17 15:54:32 +0400
@@ -93,7 +93,7 @@
 	} else {
 		cFYI(1,("dir notify on file %s Arg 0x%lx",full_path,arg));
 		rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN, 
-			GENERIC_READ | SYNCHRONIZE, 0 /* create options */,
+			GENERIC_READ | SYNCHRONIZE, FILE_SHARE_ALL, 0 /* create options */,
 			&netfid, &oplock,NULL, cifs_sb->local_nls,
 			cifs_sb->mnt_cifs_flags & CIFS_MOUNT_MAP_SPECIAL_CHR);
 		/* BB fixme - add this handle to a notify handle list */

diff -u cifs-orig148/file.c cifs/file.c
--- cifs-orig148/file.c	2007-04-05 02:11:56 +0400
+++ cifs/file.c	2007-06-05 13:37:31 +0400
@@ -106,6 +106,14 @@
 		return FILE_OPEN;
 }
 
+static inline int cifs_get_share_flags(unsigned int flags)
+{
+	if (!etersoft_flag)
+		return FILE_SHARE_ALL;
+	return ((~(flags>>29))&7);
+}
+
+
 /* all arguments to this function must be checked for validity in caller */
 static inline int cifs_open_inode_helper(struct inode *inode, struct file *file,
 	struct cifsInodeInfo *pCifsInode, struct cifsFileInfo *pCifsFile,
@@ -190,6 +196,7 @@
 	struct list_head *tmp;
 	char *full_path = NULL;
 	int desiredAccess;
+	int desiredShare;
 	int disposition;
 	__u16 netfid;
 	FILE_ALL_INFO *buf = NULL;
@@ -238,6 +245,7 @@
 	cFYI(1, (" inode = 0x%p file flags are 0x%x for %s",
 		 inode, file->f_flags, full_path));
 	desiredAccess = cifs_convert_flags(file->f_flags);
+	desiredShare = cifs_get_share_flags(file->f_flags);
 
 /*********************************************************************
  *  open flag mapping table:
@@ -287,7 +295,7 @@
 
 	if (cifs_sb->tcon->ses->capabilities & CAP_NT_SMBS)
 		rc = CIFSSMBOpen(xid, pTcon, full_path, disposition, 
-			 desiredAccess, CREATE_NOT_DIR, &netfid, &oplock, buf,
+			 desiredAccess, desiredShare, CREATE_NOT_DIR, &netfid, &oplock, buf,
 			 cifs_sb->local_nls, cifs_sb->mnt_cifs_flags
 				 & CIFS_MOUNT_MAP_SPECIAL_CHR);
 	else
@@ -371,6 +379,7 @@
 	struct cifsInodeInfo *pCifsInode;
 	char *full_path = NULL;
 	int desiredAccess;
+	int desiredShare;
 	int disposition = FILE_OPEN;
 	__u16 netfid;
 
@@ -411,6 +420,7 @@
 	cFYI(1, (" inode = 0x%p file flags are 0x%x for %s",
 		 inode, file->f_flags,full_path));
 	desiredAccess = cifs_convert_flags(file->f_flags);
+	desiredShare = cifs_get_share_flags(file->f_flags);
 
 	if (oplockEnabled)
 		oplock = REQ_OPLOCK;
@@ -430,7 +440,7 @@
 		FreeXid(xid);
 		return -ENOMEM;
 	} */
-	rc = CIFSSMBOpen(xid, pTcon, full_path, disposition, desiredAccess,
+	rc = CIFSSMBOpen(xid, pTcon, full_path, disposition, desiredAccess, desiredShare,
 			 CREATE_NOT_DIR, &netfid, &oplock, NULL,
 			 cifs_sb->local_nls, cifs_sb->mnt_cifs_flags & 
 				CIFS_MOUNT_MAP_SPECIAL_CHR);

diff -u cifs-orig148/inode.c cifs/inode.c
--- cifs-orig148/inode.c	2007-04-05 02:11:56 +0400
+++ cifs/inode.c	2007-05-17 15:54:32 +0400
@@ -231,7 +231,7 @@
 		return -EINVAL;	 /* EOPNOTSUPP? */
 	}
 		
-	rc = CIFSSMBOpen(xid, pTcon, path, FILE_OPEN, GENERIC_READ,
+	rc = CIFSSMBOpen(xid, pTcon, path, FILE_OPEN, GENERIC_READ, FILE_SHARE_ALL,
 			 CREATE_NOT_DIR, &netfid, &oplock, NULL,
 			 cifs_sb->local_nls,
 			 cifs_sb->mnt_cifs_flags &
@@ -629,8 +629,8 @@
 		int oplock = FALSE;
 		__u16 netfid;
 
-		rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN, DELETE,
-				 CREATE_NOT_DIR | CREATE_DELETE_ON_CLOSE,
+		rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN, DELETE, FILE_SHARE_ALL,
+				 CREATE_NOT_DIR | CREATE_DELETE_ON_CLOSE, 
 				 &netfid, &oplock, NULL, cifs_sb->local_nls,
 				 cifs_sb->mnt_cifs_flags & 
 					CIFS_MOUNT_MAP_SPECIAL_CHR);
@@ -674,7 +674,7 @@
 			   and pass in pid of opener to function */
 				rc = CIFSSMBOpen(xid, pTcon, full_path,
 						 FILE_OPEN, SYNCHRONIZE |
-						 FILE_WRITE_ATTRIBUTES, 0,
+						 FILE_WRITE_ATTRIBUTES, FILE_SHARE_ALL, 0,
 						 &netfid, &oplock, NULL,
 						 cifs_sb->local_nls,
 						 cifs_sb->mnt_cifs_flags & 
@@ -701,7 +701,7 @@
 				__u16 netfid;
 
 				rc = CIFSSMBOpen(xid, pTcon, full_path,
-						 FILE_OPEN, DELETE,
+						 FILE_OPEN, DELETE, FILE_SHARE_ALL,
 						 CREATE_NOT_DIR |
 						 CREATE_DELETE_ON_CLOSE,
 						 &netfid, &oplock, NULL,
@@ -963,7 +963,7 @@
 		/* if renaming directory - we should not say CREATE_NOT_DIR,
 		   need to test renaming open directory, also GENERIC_READ
 		   might not right be right access to request */
-		rc = CIFSSMBOpen(xid, pTcon, fromName, FILE_OPEN, GENERIC_READ,
+		rc = CIFSSMBOpen(xid, pTcon, fromName, FILE_OPEN, GENERIC_READ, FILE_SHARE_ALL, 
 				 CREATE_NOT_DIR, &netfid, &oplock, NULL,
 				 cifs_sb_source->local_nls, 
 				 cifs_sb_source->mnt_cifs_flags & 
@@ -1428,7 +1428,7 @@
 			/* BB we could scan to see if we already have it open
 			   and pass in pid of opener to function */
 			rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN,
-					 SYNCHRONIZE | FILE_WRITE_ATTRIBUTES,
+					 SYNCHRONIZE | FILE_WRITE_ATTRIBUTES, FILE_SHARE_ALL,
 					 CREATE_NOT_DIR, &netfid, &oplock,
 					 NULL, cifs_sb->local_nls,
 					 cifs_sb->mnt_cifs_flags &

diff -u cifs-orig148/link.c cifs/link.c
--- cifs-orig148/link.c	2007-04-05 02:11:56 +0400
+++ cifs/link.c	2007-05-17 15:56:02 +0400
@@ -289,7 +289,7 @@
 		/* add open and read as in fs/cifs/inode.c */
 	
 	} else {
-		rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN, GENERIC_READ,
+		rc = CIFSSMBOpen(xid, pTcon, full_path, FILE_OPEN, GENERIC_READ, FILE_SHARE_ALL,
 				OPEN_REPARSE_POINT,&fid, &oplock, NULL, 
 				cifs_sb->local_nls, 
 				cifs_sb->mnt_cifs_flags & 


Fix detecting overlaps completely (use end??)
--- cifs-orig148/file.c	2007-04-05 02:11:56 +0400
+++ cifs/file.c	2007-06-11 19:40:36 +0400
@@ -788,7 +804,7 @@
 			down(&fid->lock_sem);
 			list_for_each_entry_safe(li, tmp, &fid->llist, llist) {
 				if (pfLock->fl_start <= li->offset &&
-						length >= li->length) {
+						pfLock->fl_start+length >= li->offset+li->length) {
 					stored_rc = CIFSSMBLock(xid, pTcon, netfid,
 							li->length, li->offset,
 							1, 0, li->type, FALSE);


Force revalidate for inode
--- cifs/cifsfs.c 2007-06-12 22:05:31 +0400
+++ cifs/cifsfs.c 2007-06-14 18:44:53 +0400
@@ -590,6 +590,19 @@
 	return written;
 }
 
+static ssize_t cifs_file_read(struct file *file, char *user, size_t cnt, loff_t *pos)
+{
+	if( file!=NULL && file->f_dentry!=NULL && CIFS_I(file->f_dentry->d_inode)!=NULL ) {
+		int retval = 0;
+		CIFS_I(file->f_dentry->d_inode)->needForceInvalidate = 1;
+		retval = cifs_revalidate(file->f_dentry);
+		if( retval < 0 )
+			return (ssize_t)retval;
+	}
+
+	return do_sync_read(file,user,cnt,pos);
+}
+
 static loff_t cifs_llseek(struct file *file, loff_t offset, int origin)
 {
 	/* origin == SEEK_END => we must revalidate the cached file length */
@@ -679,7 +692,7 @@
 };
 
 const struct file_operations cifs_file_ops = {
-	.read = do_sync_read,
+	.read = cifs_file_read,
 	.write = do_sync_write,
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 	.readv = generic_file_readv,
@@ -989,6 +1002,10 @@
 				to server still is disconnected since oplock 
 				already released by the server in that case */
 				if(pTcon->tidStatus != CifsNeedReconnect) {
+					/* PV: disable caching if oplock missed  */
+					CIFS_I(inode)->clientCanCacheRead = FALSE;
+					CIFS_I(inode)->clientCanCacheAll = FALSE;
+
 				    rc = CIFSSMBLock(0, pTcon, netfid,
 					    0 /* len */ , 0 /* offset */, 0, 
 					    0, LOCKING_ANDX_OPLOCK_RELEASE,


--- cifs/inode.c	2007-06-12 22:05:31 +0400
+++ cifs/inode.c	2007-06-14 21:42:57 +0400
@@ -1027,7 +1027,7 @@
 		 direntry->d_inode->i_count.counter, direntry,
 		 direntry->d_time, jiffies));
 
-	if (cifsInode->time == 0) {
+	if (cifsInode->time == 0 || cifsInode->needForceInvalidate ) {
 		/* was set to zero previously to force revalidate */
 	} else if (time_before(jiffies, cifsInode->time + HZ) &&
 		   lookupCacheEnabled) {
@@ -1069,18 +1069,18 @@
 	/* if not oplocked, we invalidate inode pages if mtime or file size
 	   had changed on server */
 
-	if (timespec_equal(&local_mtime,&direntry->d_inode->i_mtime) && 
-	    (local_size == direntry->d_inode->i_size)) {
-		cFYI(1, ("cifs_revalidate - inode unchanged"));
+	if (!cifsInode->needForceInvalidate &&
+		timespec_equal(&local_mtime,&direntry->d_inode->i_mtime) && 
+	    (local_size == direntry->d_inode->i_size) ) {
+		cFYI(1, ("***************************** cifs_revalidate - inode unchanged"));
 	} else {
 		/* file may have changed on server */
 		if (cifsInode->clientCanCacheRead) {
 			/* no need to invalidate inode pages since we were the
 			   only ones who could have modified the file and the
 			   server copy is staler than ours */
-		} else {
+		} else
 			invalidate_inode = TRUE;
-		}
 	}
 
 	/* can not grab this sem since kernel filesys locking documentation
@@ -1099,10 +1099,17 @@
 /*		if(S_ISDIR(direntry->d_inode->i_mode)) 
 			shrink_dcache_parent(direntry); */
 		if (S_ISREG(direntry->d_inode->i_mode)) {
+			/* (pv): Зачем здесь второй вызов filemap_fdatawait
+				Ведь мы выше уже вызывали эту функцию. */
 			if (direntry->d_inode->i_mapping)
 				filemap_fdatawait(direntry->d_inode->i_mapping);
+
+			if( cifsInode->needForceInvalidate ) {
+				cFYI(1, ("Force invalidating."));
+ 				invalidate_remote_inode(direntry->d_inode);
+				cifsInode->needForceInvalidate = 0;
 			/* may eventually have to do this for open files too */
-			if (list_empty(&(cifsInode->openFileList))) {
+			} else if (list_empty(&(cifsInode->openFileList))) {
 				/* changed on server - flush read ahead pages */
 				cFYI(1, ("Invalidating read ahead data on "
 					 "closed file"));


--- cifs/cifsglob.h	2007-06-12 22:05:31 +0400
+++ cifs/cifsglob.h	2007-06-14 16:04:07 +0400
@@ -345,6 +345,7 @@
 	unsigned clientCanCacheRead:1; /* read oplock */
 	unsigned clientCanCacheAll:1;  /* read and writebehind oplock */
 	unsigned oplockPending:1;
+	unsigned needForceInvalidate:1; 
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2, 5, 0)
 	struct inode vfs_inode;
 #endif

Correct build for ovz-kernels
Надо по этому патчу внести изменения в исходный код, чтобы условие было на наличие utsname, а не на версию ядра
--- cifs/connect.c	2007-06-17 22:44:40 +0400
+++ cifs/connect.c	2007-06-17 23:01:57 +0400
@@ -39,6 +39,7 @@
 #endif
 #include <asm/uaccess.h>
 #include <asm/processor.h>
+#include <linux/autoconf.h>
 #include "cifspdu.h"
 #include "cifsglob.h"
 #include "cifsproto.h"
@@ -826,7 +827,7 @@
 	separator[1] = 0; 
 
 	memset(vol->source_rfc1001_name,0x20,15);
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 	for(i=0;i < strnlen(utsname()->nodename,15);i++) {
 		/* does not have to be a perfect mapping since the field is
 		informational, only used for servers that do not support
@@ -2248,7 +2249,7 @@
 				  32, nls_codepage);
 		bcc_ptr += 2 * bytes_returned;
 		bytes_returned =
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 			cifs_strtoUCS((__le16 *)bcc_ptr, utsname()->release,
                                   32, nls_codepage);
 
@@ -2281,7 +2282,7 @@
 		}
 		strcpy(bcc_ptr, "Linux version ");
 		bcc_ptr += strlen("Linux version ");
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
                 strcpy(bcc_ptr, utsname()->release);
                 bcc_ptr += strlen(utsname()->release) + 1;
 #else
@@ -2551,7 +2552,7 @@
 				  32, nls_codepage);
 		bcc_ptr += 2 * bytes_returned;
 		bytes_returned =
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 		    cifs_strtoUCS((__le16 *)bcc_ptr, utsname()->release, 32,
 				nls_codepage);
 
@@ -2574,7 +2575,7 @@
 	} else {		/* ASCII */
 		strcpy(bcc_ptr, "Linux version ");
 		bcc_ptr += strlen("Linux version ");
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 		strcpy(bcc_ptr, utsname()->release);
 		bcc_ptr += strlen(utsname()->release) + 1;
 #else
@@ -2954,7 +2955,7 @@
 		bcc_ptr += 2 * bytes_returned;
 		bytes_returned =
 
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 		    cifs_strtoUCS((__le16 *) bcc_ptr, utsname()->release, 32,
                                   nls_codepage);
 #else
@@ -3011,7 +3012,7 @@
 
 		strcpy(bcc_ptr, "Linux version ");
 		bcc_ptr += strlen("Linux version ");
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 		strcpy(bcc_ptr, utsname()->release);
 		bcc_ptr += strlen(utsname()->release) + 1;
 #else

--- cifs/sess.c.orig	2007-06-17 22:44:45 +0400
+++ cifs/sess.c	2007-06-17 23:02:33 +0400
@@ -30,6 +30,7 @@
 #include "ntlmssp.h"
 #include "nterr.h"
 #include <linux/utsname.h>
+#include <linux/autoconf.h>
 
 extern void SMBNTencrypt(unsigned char *passwd, unsigned char *c8,
                          unsigned char *p24);
@@ -117,7 +118,7 @@
 	bytes_ret = cifs_strtoUCS((__le16 *)bcc_ptr, "Linux version ", 32,
 				  nls_cp);
 	bcc_ptr += 2 * bytes_ret;
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 	bytes_ret = cifs_strtoUCS((__le16 *) bcc_ptr, utsname()->release,
 				  32, nls_cp);
 #else
@@ -167,7 +168,7 @@
 
 	strcpy(bcc_ptr, "Linux version ");
 	bcc_ptr += strlen("Linux version ");
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 18) || defined CONFIG_VE
 	strcpy(bcc_ptr, utsname()->release);
 	bcc_ptr += strlen(utsname()->release) + 1;
 #else

Fix compiling for 2.6.9 and older
--- cifs/inode.c.orig	2007-06-12 22:05:31 +0400
+++ cifs/inode.c	2007-06-18 23:18:59 +0400
@@ -1178,7 +1185,11 @@
 	goto out_truncate;
 
 do_expand:
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,9)
 	limit = current->signal->rlim[RLIMIT_FSIZE].rlim_cur;
+#else
+	limit = current->rlim[RLIMIT_FSIZE].rlim_cur;
+#endif
 	if (limit != RLIM_INFINITY && offset > limit) {
 		spin_unlock(&inode->i_lock);
 		goto out_sig;

Fin compiling for 2.6.18 in Fedora 7 and CentOS 5
--- cifs/cifsfs.c.orig	2007-06-19 00:05:21 +0400
+++ cifs/cifsfs.c	2007-06-20 23:13:10 +0400
@@ -294,7 +294,7 @@
 	file data or metadata */
 	cifs_inode->clientCanCacheRead = FALSE;
 	cifs_inode->clientCanCacheAll = FALSE;
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,19)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18)
 	cifs_inode->vfs_inode.i_blksize = CIFS_MAX_MSGSIZE;
 #endif
 	cifs_inode->vfs_inode.i_blkbits = 14;  /* 2**14 = CIFS_MAX_MSGSIZE */

--- cifs/asn1.c.orig	2007-06-22 22:53:02 +0400
+++ cifs/asn1.c	2007-06-22 22:53:05 +0400
@@ -26,7 +26,7 @@
 #include "cifsglob.h"
 #include "cifs_debug.h"
 #include "cifsproto.h"
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,19)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18)
 #include <linux/config.h>
 #endif
 

--- cifs-smp-o/cifsfs.c	2007-06-05 12:26:04 +0400
+++ cifs/cifsfs.c	2007-07-03 11:40:34 +0400
@@ -184,7 +184,7 @@
 }
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 16)
-void * kzalloc(size_t size, unsigned flgs)
+void * kzalloc(size_t size, int flgs)
 {
 	void * buf;
 	buf = kmalloc(size, flgs);

--- cifs-smp-o/cifsproto.h	2007-05-17 15:54:32 +0400
+++ cifs/cifsproto.h	2007-07-03 11:40:19 +0400
@@ -366,6 +366,6 @@
 extern int CIFSGetExtAttr(const int xid, struct cifsTconInfo *tcon,
                 const int netfid, __u64 * pExtAttrBits, __u64 *pMask);
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 16)
-extern void * kzalloc(size_t size, unsigned flgs);
+extern void * kzalloc(size_t size, int flgs);
 #endif
 #endif			/* _CIFSPROTO_H */


В 2.6.22 их больше нет?
--- cifs/cifsfs.c.orig	2007-04-05 02:11:56 +0400
+++ cifs/cifsfs.c	2007-07-27 20:23:34 +0400
@@ -782,11 +782,8 @@
 {
 	struct cifsInodeInfo *cifsi = inode;
 
-	if ((flags & (SLAB_CTOR_VERIFY | SLAB_CTOR_CONSTRUCTOR)) ==
-	    SLAB_CTOR_CONSTRUCTOR) {
-		inode_init_once(&cifsi->vfs_inode);
-		INIT_LIST_HEAD(&cifsi->lockList);
-	}
+	inode_init_once(&cifsi->vfs_inode);
+	INIT_LIST_HEAD(&cifsi->lockList);
 }
 
 static int
